utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -289,31 +281,47 @@ private void doProcessQueue() {
                 entry.resultFuture.complete(null);
             }
 
-            eventQueue.remove();
+            sourceQueue.remove();
+        }
+    }
+
+    private void trySetFailureMessage(Supplier<Throwable> supplier) {
+        if (failureMessageSupplier == null) {
+            failureMessageSupplier = supplier;
         }
     }
 
+    private boolean hasFailureMessage() {
```

Comment:
This method is  bit confusing since we are checking         ` failureMessageSupplier != null;` while we have failureMessage field

Reply:
I can see why that would be confusing. All of the methods relate to setting and checking the failure message. The implementation uses a supplier (failureMessageSupplier) and cached result (failureMessage), but that's an implementation detail of those methods that the caller shouldn't have to worry about, so I named the methods accordingly.

I can move those fields and methods into an inner class so that it's clear those fields are encapsulated to the failure-related methods.